kernel/src/scan/mod.rs

```diff
@@ -311,11 +316,13 @@ impl Scan {
                         mask: sv,
                     };
                     selection_vector = rest;
-                    Ok(result)
+                    Ok::<ScanResult, Error>(result)
```

Comment:
On line 296 above you can help the compiler by telling what type the closure will return:
```rust
                Ok(read_result_iter.into_iter().map(move |read_result| -> DeltaResult<_> {
```

Then you don't need a turbofish here:
```suggestion
                    Ok(result)
```

Reply:
Any suggestions to wrap my head around these types? The return type of this closure (according to my LSP) is `DeltaResult<Map<Box<dyn Iterator<Item = Result<Box<dyn EngineData>, Error>> + Send>, impl FnMut(Result<Box<dyn EngineData>, Error>) -> Result<ScanResult, Error>>>
` 😅. I'd like to do a `DeltaResult<Iterator<Item = DeltaResult<ScanResult>>>`, but that yields a pretty ugly compiler error.

Reply:
The types are super complex because of all the nested iterators and so on.  Essentially we have 3 layers of iterators:
1. The outer iterator: `scan_data`, each element here is basically one file in the log, so a `0000000..0002.json` for example
2.  We iterate over each chunk of data, which can contain multiple `add` actions each, and produce a `scan_files_iter` which tells us which files we need to scan
3. Each scan file we read produces an iterator because the parquet reader is allowed to return things in "chunks"

So ultimately we have an iterator inside an iterator inside an iterator, so we need to flatten it out three time. I don't even bother trying to unravel the types because most of them are actually adapters (like the return from a `map` say, and it's better to just think of them as "just another iterator".

Regarding the turbofish and the type annotation. Say you have an iterator over results, and you want to map to another result, you can do something like:

```rust
some_iter.map(|result| {
  let thing = result?.do_something();
  let other_thing = fallible_call()?;
  Ok(thing)
})
```

but now the compiler needs to know what the error type is. `result` might be a `Result<_, ErrorOne>` and `fallible_call()` might return `Result<_, ErrorTwo>`. We have `into`s for everything to go into a `delta_kernel::Error` type, but the compiler doesn't know that's what we want it to use when we do `?`, so it complains. You can fix this by saying `Ok::<_, delta_kernel::Error>` which is just telling it fully what the return type is. But turbofish is ugly, so we use the _other_ way to tell it what the return type is by adding a type annotation to the closure:
```rust
some_iter.map(|result| -> Result<_, UnifiedErrorType> {
  let thing = result?.do_something();
  let other_thing = fallible_call()?;
  Ok(thing)
})
```

Reply:
That helps a lot, thank you!